src/programming-assistants.rst

```diff
@@ -1,23 +1,179 @@
+
+1. Complete Formal Specification using a formula (e.g. `∀x. P(x) = x*2` for a program that doubles the input)
+2. Input/Output pairs that serve as examples of what the program should be doing (e.g. (2,4), (4,8) for the program that doubles)
+3. Demonstrations of how the program should compute the output (Very similar to #2 but tends to have some intermediate steps as well)
+
+The two main types of Program Synthesis are deductive and inductive. **Deductive Program Synthesis** uses complete formal specifications while **Inductive Program Synthesis** uses less formal specifications and relies primarily on iterative search techniques to find implementations
+
+Deductive Program Synthesis is focused on an approach in which the derivation task is regarded as a problem of proving a mathematical theorem :cite:`manna1992fundamentals`. Manna and Waldinger discuss that in order to write a program that meets the specifications, a proof of existence must be conducted showing that for any input, the output satisfies the specified conditions. In order for the proof to be sufficiently constructive, upon finding the existence of an output, it is required to state the computational means used to determine the existence. This is then used as a basis for the program that is eventually created. Manna and Waldinger continue to discuss specifications, proving of theorems and deduction rules for the remainder of the paper.
+
+Inductive Program Synthesis (IPS) has been a more popular area of research in the last decade. This is mainly due to the realization that most are not familiar with complicated mathematical theorems and prefer to synthesize using less formal specifications. There are many different variations of IPS as there are many different iterative approaches, but most follow the same overall architecture containing a synthesizer and a verifier:
+
+.. image:: program_synthesis_model.png
+
+A synthesizer produces a 'candidate program' based on the specifications provided and sends the program to a verifier to validate that the program fully satisfies the specifications. If the verifier successfully validates the program, no further action is taken. However, if the verifier determines that the candidate is not satisfactory, it provides feedback to the synthesizer to be used to guide the development of a new candidate program. This process is repeated until the candidate program passes all of the specification.
+
+Gvero and Kuncak have a proposed a new code assistance tool, anyCode, for integrated development environments (IDEs) that produces a list of syntheses based on user-entered free-form queries :cite:`gvero2015synthesizing`. The queries contain a mixture of English and Java syntax and produce expressions that account for the correct syntax, types and scoping rules of Java.
```

Comment:
Put this in the summary for the corresponding bib item below (it's okay if it is long)

Reply:
Should there be no text besides the citation then?